### PR TITLE
SW-695 Use Shopware category tree for cross-selling-category selection

### PR DESCRIPTION
**Pull request by TobiasGraml11**

### CODE REVIEW

File: src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js [hunk 1]
```diff
@@ -54,19 +59,17 @@ Component.register('findologic-config', {
   data () {
     return {
       isLoading: false,
-      term: null,
-      categories: [],
-      categoryIds: []
+      categoryCollection: null,
```
Comment (TobiasGraml11) [inline src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js]: The concept of this implementation was taken from https://github.com/shopware/platform/blob/trunk/src/Administration/Resources/app/administration/src/module/sw-sales-channel/view/sw-sales-channel-detail-base/index.js

Comment (dblaichinger) [reply src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js]: Really great change, it makes the code less complex and the category select in the same way as Shopware intended it 👍

File: src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js [hunk 2]
```diff
@@ -229,16 +194,15 @@ Component.register('findologic-config', {
       return this.repositoryFactory.create('category');
     },
 
-    categoryCriteria () {
-      const criteria = new Criteria(1, 500);
-      criteria.addSorting(Criteria.sort('name', 'ASC'));
-      criteria.addSorting(Criteria.sort('parentId', 'ASC'));
+    categories () {
+      return this.categoryCollection ? this.categoryCollection : [];
```
Comment (dblaichinger) [inline src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js]: `??` could be used.

**Review by dblaichinger — CHANGES_REQUESTED** [review]

File: src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js [hunk 3]
```diff
@@ -146,29 +126,14 @@ Component.register('findologic-config', {
...
+    },
 
-        this.categories = this.sortByProperty(translatedCategories, 'label');
-      }).finally(() => {
-        this.isLoading = false;
-      });
-    }
+    onCategoryRemove(item) {
+      this.actualConfigData['FinSearch.config.crossSellingCategories'] =
+        this.actualConfigData['FinSearch.config.crossSellingCategories'].filter(categoryId => categoryId !== item.id);
```
Comment (TheKeymaster) [inline src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js:133]: `.filter()` goes though all elements. Wouldn't it be easier to just pop the element of the array?
```suggestion
        delete this.actualConfigData['FinSearch.config.crossSellingCategories'][item.id];
```

File: src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js [hunk 4]
```diff
@@ -54,19 +54,25 @@ Component.register('findologic-config', {
...
-      this.getCategories();
+    async createCategoryCollection () {
+      this.categoryCollection = this.actualConfigData['FinSearch.config.crossSellingCategories'].length
+        ? await this.categoryRepository.search(this.selectedCategoriesCriteria, Shopware.Context.api)
+        : new EntityCollection(
+          this.categoryRepository.route,
+          this.categoryRepository.entityName,
+          Shopware.Context.api,
+          null,
+          []
```
Comment (TheKeymaster) [inline src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js]: `criteria` and `entities` already should have sufficient defaults.

image

File: src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js [hunk 5]
```diff
@@ -229,16 +194,15 @@ Component.register('findologic-config', {
       return this.repositoryFactory.create('category');
     },
 
-    categoryCriteria () {
-      const criteria = new Criteria(1, 500);
-      criteria.addSorting(Criteria.sort('name', 'ASC'));
-      criteria.addSorting(Criteria.sort('parentId', 'ASC'));
+    categories () {
+      return this.categoryCollection ?? [];
```
Comment (TheKeymaster) [inline src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js]: Why does this even exist? Can't `this.categoryCollection` already be `[]` by default?

File: src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js [hunk 6]
```diff
@@ -146,29 +126,14 @@ Component.register('findologic-config', {
...
-        this.term = null;
-        this.total = items.total;
-        items.forEach((category) => {
-          translatedCategories.push({
-            value: category.id,
-            name: category.name,
-            label: category.translated.breadcrumb.join(' > ')
-          });
-        });
+    onCategoryAdd(item) {
```
Comment (TheKeymaster) [inline src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js:127]: Does the `item` has any API reference that we could add as a docblock? Would be easier for debugging when knowing how the structure looks like.

**Review by TheKeymaster — CHANGES_REQUESTED** [review]

File: src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js (hunk 3 above)
Comment (TobiasGraml11) [reply src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js:133]: The array is not indexed by the item id:

image

File: src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js (hunk 5 above)
Comment (TobiasGraml11) [reply src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js]: I am not sure about it, I am just using the logic from Shopware.

File: src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js (hunk 6 above)
Comment (TobiasGraml11) [reply src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js:127]: This component event is not documented at all: https://component-library.shopware.com/components/sw-category-tree-field

Just saw it in the implementation of Shopware. It should be a category entity.

File: src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js (hunk 5 above)
Comment (TobiasGraml11) [reply src/Resources/app/administration/src/module/findologic-module/components/findologic-config/index.js]: I adapted it, and it seems to still work

**Review by TheKeymaster — APPROVED** [review]

**Review by dblaichinger — APPROVED** [review]

**Pull request merged**